src/psp/guest_request.rs

```diff
@@ -0,0 +1,531 @@
+type EvpCipherCtxPtr = *mut EVP_CIPHER_CTX;
+
+impl SnpGuestRequestMsg {
+    fn alloc_raw() -> SnpGuestRequestMsgPtr {
+        let pa: PhysFrame = match mem_allocate_frame() {
+            Some(f) => f,
+            None => return ptr::null_mut(),
+        };
+        let va: VirtAddr = pgtable_pa_to_va(pa.start_address());
+        pgtable_make_pages_shared(va, PAGE_SIZE);
+        memset(va.as_mut_ptr(), 0, PAGE_SIZE as usize);
+        va.as_mut_ptr() as SnpGuestRequestMsgPtr
+    }
+    fn free_raw(p: SnpGuestRequestMsgPtr) {
+        let va: VirtAddr = VirtAddr::from_ptr(p);
+        pgtable_make_pages_private(va, PAGE_SIZE);
+        free_page(va);
+    }
+}
+
+static SEQ_NUM: AtomicU64 = AtomicU64::new(0);
+fn seqno_last_used() -> u64 {
+    SEQ_NUM.load(Ordering::Relaxed)
+}
+fn seqno_add_two() {
+    SEQ_NUM.fetch_add(2, Ordering::Relaxed);
+}
+
+struct SnpGuestRequestCmd {
+    // The Guest Request NAE event requires two unique pages: one for
+    // the request and another for the response message. Both of them
+    // are assigned to the hypervisor (shared).
+    // We encrypt the request payload using AES-256-GCM and the vmpck[0]
+    // key before sending it. The PSP firmware does the same with the
+    // response payload.
+    req: SnpGuestRequestMsgPtr,
+    resp: SnpGuestRequestMsgPtr,
+    ctx_req: EvpCipherCtxPtr,
+    ctx_resp: EvpCipherCtxPtr,
+}
+
+// The SnpGuestRequestCmd fields are raw pointers and we need to properly
+// free them.
+impl Drop for SnpGuestRequestCmd {
+    fn drop(&mut self) {
+        self.free();
+    }
+}
+
+impl SnpGuestRequestCmd {
+    fn new() -> Self {
+        SnpGuestRequestCmd {
+            req: ptr::null_mut(),
+            resp: ptr::null_mut(),
+            ctx_req: ptr::null_mut(),
+            ctx_resp: ptr::null_mut(),
+        }
+    }
+
+    fn init(&mut self) -> bool {
+        self.req = SnpGuestRequestMsg::alloc_raw();
+        if self.req.is_null() {
+            prints!("ERROR: alloc_raw failed for the request\n");
+            return false;
+        }
+        self.resp = SnpGuestRequestMsg::alloc_raw();
+        if self.resp.is_null() {
+            prints!("ERROR: alloc_raw failed for the response\n");
+            return false;
+        }
+        true
+    }
+
+    fn free(&self) {
+        if !self.req.is_null() {
+            SnpGuestRequestMsg::free_raw(self.req);
+        }
+        if !self.resp.is_null() {
+            SnpGuestRequestMsg::free_raw(self.resp);
+        }
+        if !self.ctx_req.is_null() {
+            unsafe { EVP_CIPHER_CTX_free(self.ctx_req) };
+        }
+        if !self.ctx_resp.is_null() {
+            unsafe { EVP_CIPHER_CTX_free(self.ctx_resp) };
+        }
+    }
+
+    fn encrypt_request(&mut self, msg_type: u8, plaintext: &[u8]) -> bool {
+        let msg_seqno: u64 = match seqno_last_used().checked_add(1) {
+            Some(v) => v,
+            None => {
+                prints!("ERROR: Request sequence number overflow\n");
```

Comment:
Any error that causes an issue with the sequence number requires clearing of the vmpck0 key so that it can't be used again.